Handle multiple new lines

camillebaldock opened the pull request:
Creates the concept of a `MarkdownLine` and a `MarkdownBlankLine`. 
We now treat newlines as something to parse separately from a paragraph, and no longer as an ignorable delimiter. 

Introduction of Jekyll-style delimiters for front matter for ease of parsing. 

Please come talk to me, not hugely happy with the use of 'blocks' in the parsing logic, but could not figure out a way to phrase the grammar without parslet stalling...

camillebaldock committed:
Failing test for multiple new lines parsing

camillebaldock committed:
Acknowledge blank lines during parsing

camillebaldock committed:
Use Jekyll-style delimiters for front matter

jackscotti commented on lib/smartdown/model/block.rb:
```diff
@@ -0,0 +1,5 @@
+  module Model
+    Block = Struct.new(:elements) 
+  end
+end
```
A new line please :)

camilleldn commented:
Code Review pair: 
* renaming of blocks as Elements
* stop using MarkdownBlankLine, revert to just MarkdownLine

camillebaldock committed:
Refactor Markdown blank line parsing

Stop using a separate object for blank lines: they are just a line
Rename the "block" parsing element to be "elements": more descriptive
of what that object contains

We had to create this Struct for storing multiple Smartdown::Model elements due
to the syntax used in parsing.
In the parsing rule below:
```
rule(:markdown_elements) {
  markdown_element.repeat(1, 1) >>
  (newline.repeat(1).as(:blank_line) >> markdown_element.as(:element)).repeat
}
```
the `(newline.repeat(1).as(:blank_line) >> markdown_element.as(:element)).repeat` snippet creates
a hash with two key/value pairs to transform.
In order to avoid too much verbosity in the NodeTransform, using the markup "element" instead
of relying on the naming given by the markdown element parsing itself means we can catch all those cases
in one generic transform rule:
```
rule(blank_line: simple(:blank_line), element: subtree(:element)) {
  Smartdown::Model::Elements.new(
    [
      Smartdown::Model::Element::MarkdownLine.new(blank_line.to_s),
      element,
    ]
  )
}
```
The other alternative would have been to have to add a new rule for every specific kind of markdown element.
This `Elements` Struct was created to avoid that much verbosity and brittleness in the transform (we would
have needed to add another `rule (blank_line...., xxxxx....)` every time we add a new type of markdown element)
Last but not least, we couldn't just cast newlines as just any markdown element:
```
rule(:markdown_elements) {
  markdown_element.repeat(1, 1) >>
  (markdown_element).repeat
}
```
was not an option since we needed to differentiate between newlines that so happened to be inside broader blocks like
frontmatter, rules or conditionals, and newlines that separated plain markdown text...

dsingleton commented on lib/smartdown/model/node.rb:
```diff
@@ -4,7 +4,14 @@ module Smartdown
     Node = Struct.new(:name, :elements, :front_matter) do
       def initialize(name, elements, front_matter = nil)
-        super(name, elements, front_matter || Smartdown::Model::FrontMatter.new)
+        all_elements = elements.map do |element|
```
I'd be good to have an inline comment here (and/or in the `Elements` class itself) explaining why there is the `Elements` model that needs flattening. We don't want too much information about how the parser works to leak into other layers, but we should explain it somewhere. What do you think?

dsingleton commented on lib/smartdown/model/node.rb:
```diff
@@ -73,8 +80,10 @@ def elements_of_kind(*kinds)
+          "# #{block.content}"
+        when Smartdown::Model::Element::MarkdownLine
+          block.content
+        when Smartdown::Model::Element::MarkdownBlankLine
```
`MarkdownBlankLine` is gone now right?

dsingleton commented:
Looks good to me, pending a couple of comments i'd be happy to merge :+1:
